Initial version for supporting the IMC-2001 (Taiwanese Apple ][ clone)

andreasbaumann opened the pull request:
* a new entry in the list of clones and a dedicated RUN button
* has differences in IMC2001.rom to support Apple ][e like lowercase characters in the monitor ROM
* the video ROM for 40-character more IMC2001-Video.rom is quite different (mainly to support lowercase ASCII characters in NORMAL mode)
* the keybaord act more like a Apple][e, decided to use the CAPS mode switch as for the Apple][e in the emulator
* TODO: support for special keyboard features like the separate HOME, DEL, UP, DOWN, LEFT, RIGHT block of keys. Also F1 to F12 keys which are hard-coded in the keyboard ROM (with a switch for APPLE and CP/M mode)

andreasbaumann commented:
Some pictures, documentation and background information is on https://www.andreasbaumann.cc/hardware/imc-2001/

tomcw commented on source/Keyboard.cpp:
```diff
@@ -223,6 +223,14 @@ void KeybQueueKeypress (WPARAM key, Keystroke_e bASCII)
+				// TODO: the IMC-2001 actually has black LEFT and RIGTH arrow keys which behave as
+				// on an Apple IIe, but it also has light arrow keys UP, DOWN, LEFT, RIGHT, which
+				// actually position the cursor on the screen, not sure how to emulate this.
+				// for now just doing the black LEFT and RIGHT keys
```
Typo: "RIGTH"
Also the sentence is incomplete "which behave as ? on an Apple IIe"

tomcw commented on source/Keyboard.cpp:
```diff
@@ -450,7 +458,7 @@ BYTE KeybReadFlag (void)
 void KeybToggleCapsLock ()
 {
-	if (!IS_APPLE2)
+	if (!IS_APPLE2 ||IsIMC2001(g_Apple2Type))
```
Missing a space after "||"

tomcw commented on source/NTSC_CharSet.cpp:
```diff
@@ -209,6 +215,12 @@ static void userVideoRom2K(csbits_t csbits, const BYTE* pVideoRom, const eApple2
+				else if (type == A2TYPE_IMC2001)
+				{
+					// for now just unconditionally negate the pattern for inverse (does
+					// this work for small letters too?)
```
formatting: move the "(does" onto the 2nd comment line

andreasbaumann replied:
Hi, sorry about the typos.

tomcw replied:
Thanks!

tomcw reviewed (changes requested):
I've eye-balled this PR, and there are a few comments regarding typos & formatting that I'd like addressed.

tomcw commented:
Can you also add new info for the IMC-2001 in the "Clone" section in the Help html file:
https://github.com/AppleWin/AppleWin/blob/master/help/cfg-advanced.html

tomcw commented:
Hi @andreasbaumann,
Thanks for submitting this PR for this Apple II clone.

There are outstanding TODO items that you have identified. Do you plan to tackle any of these? Or do you think it's "good enough" to merge this PR into the main line now?

andreasbaumann commented:
It's usable as it is now. The trouble currently is, that the original machine is in mid-
restauration (every time I power it on, some new weird stuff happens). So, it's not
that easy to verify that everything is now the same as in the emulator. :-)

I could verify a few things:
- the inverse, flash modes and also the character under the cursor are wrong also
  on the original if you write small characters (it works in normal mode only), so I'll
  document accordingly and remove the TODO as this is the behaviour of the
  IMC-2001.
- the free floating cursor I can tackle, would be neat, but no other emulation mode
  seems to do it right now, so it might need some changes in the more core part of
  the screen handling. This can also be post-poned.
- a new TODO popped up: the language card has to be always enabled, as the two
  ROMS just contain the functionality of the language card and it can not be turned
  on and off (it's always on).

There is no particular hurry to merge it now, maybe let it "ripe" for another one or
two weeks. :-)

I also adapted the Linux port (audetto), Github makes it harder than necessary to
do patches in the original and the forked version of a project. There I just added
some values to the imlib, qt frontends and added the resources. I'll make a pull
request there later when this pull request here is done - also not in a big hurry here.

There is more stuff for future pull requests:
- the motherboard has a 6502 and a Z80 on it, sharing the same RAM. Thus it
  supports CP/M on board.
- It has a Vidtex Videoterm card for 80-column mode (with a hardware switch
  diretly in the video output). It works totally different than the 80-column mode on
  the later Apple models: the video ram is bank-switched in the $C000 area and
  manuals I find are not completely describing what the hardware is doing, there are
  differences I have to identify first.

tomcw commented:
>There is no particular hurry to merge it now, maybe let it "ripe" for another one or two weeks. :-)

Yes, this is completely fine - take your time (no rush or pressure! :-)

Let's use this PR issue to discuss how to achieve certain things when they don't quite fit into AppleWin's current structure (eg. extra keys, Videx-style 80 col video, etc).

andreasbaumann commented:
On the light-colored arrow keys: I just saw on
https://www.andreasbaumann.cc/hardware/imc-2001/manual_imc_2001_2.01.pdf, page 56 that

- `ESC @` is mapped to `HOME`
- `ESC A-D` to those light arrow keys.

And the escape sequences already work in the emulator and are the standard escape sequences
on an Apple ][.

So all the IMC-2001 keyboard controller does, is, it sends keyboard sequences when you press
the arrow keys or the HOME key.

tomcw commented:
>It has a Vidtex Videoterm card for 80-column mode (with a hardware switch
diretly in the video output). It works totally different than the 80-column mode on
the later Apple models: the video ram is bank-switched in the $C000 area and
manuals I find are not completely describing what the hardware is doing, there are
differences I have to identify first.

Sounds like it'll need AppleWin to support a Videx Videoterm (80 col) card.
There's already an issue requesting this: #105
